File: src/Infrastructure/Repositories/GenericRepository.cs
using Domain.Interface;
using Infrastructure.Context;
using Microsoft.EntityFrameworkCore;

namespace Infrastructure.Repositories;

public class GenericRepository<TEntity>(RssReaderDbContext rssReaderDbContext)
    : IGenericRepository<TEntity> where TEntity : class
{
    internal readonly DbSet<TEntity> DbSet = rssReaderDbContext.Set<TEntity>();

    public async Task<TEntity> AddAsync(TEntity entity)
    {
        var entry = await DbSet.AddAsync(entity);
        return entry.Entity;
    }

    public void Delete(TEntity entity)
    {
        DbSet.Remove(entity);
    }

    public async Task<IReadOnlyList<TEntity>> GetAllAsync()
    {
        return await DbSet.ToListAsync();
    }

    public async Task<TEntity?> GetByIdAsync(int id)
    {
        return await DbSet.FindAsync(id);
    }

    public TEntity Update(TEntity entity)
    {
        var entry = DbSet.Update(entity);
        return entry.Entity;
    }
}

File: src/ClientApp/src/app/infrastructure/repositories/GenericRepository.ts
import { IGenericService } from "../../domain/interfaces/IGenericService";

export class GenericRepository<TEntity> implements IGenericService<TEntity> {
  private entities: TEntity[] = [];

  async getByIdAsync(id: number): Promise<TEntity | null> {
    const entity = this.entities.find(e => (e as any).id === id) || null;
    return await Promise.resolve(entity);
  }

  async getAllAsync(): Promise<TEntity[]> {
    return await Promise.resolve(this.entities);
  }

  async addAsync(entity: TEntity): Promise<TEntity> {
    this.entities.push(entity);
    return await Promise.resolve(entity);
  }

  update(entity: TEntity): TEntity {
    const index = this.entities.findIndex(e => (e as any).id === (entity as any).id);
    if (index !== -1) {
      this.entities[index] = entity;
    }
    return entity;
  }

  delete(entity: TEntity): void {
    this.entities = this.entities.filter(e => (e as any).id !== (entity as any).id);
  }
}



